File: internal/discovery/discovery.go
package discovery

import (
	"bytes"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/JustinDFuller-org/ban-code-comments/internal/languages"
	"github.com/JustinDFuller-org/ban-code-comments/internal/model"
)

type Config struct {
	Paths     []string
	Languages map[model.Language]bool
	Includes  []string
	Excludes  []string
	Debug     bool
}

type Candidate struct {
	Path     string
	RelPath  string
	Language model.Language
	gitRoot  string
}

type Diagnostic struct {
	Path   string
	Reason string
}

type Result struct {
	Candidates []Candidate
	Skipped    int
	Debug      []Diagnostic
}

var fixedDirectories = map[string]bool{
	".git": true, ".hg": true, ".svn": true, "node_modules": true, "vendor": true,
	"dist": true, "build": true, "target": true, ".build": true, ".next": true,
	"coverage": true, "tmp": true, ".cache": true,
}

func Discover(config Config) (Result, error) {
	workingDirectory, err := os.Getwd()
	if err != nil {
		return Result{}, err
	}
	workingDirectory, err = filepath.Abs(workingDirectory)
	if err != nil {
		return Result{}, err
	}
	if resolvedDirectory, resolveErr := filepath.EvalSymlinks(workingDirectory); resolveErr == nil {
		workingDirectory = resolvedDirectory
	}
	gitRoot := findGitRoot(workingDirectory)
	var candidates []Candidate
	seenCandidates := make(map[string]bool)
	rootCache := make(map[string]string)
	var diagnostics []Diagnostic
	skipped := 0
	addCandidate := func(candidate *Candidate) {
		if seenCandidates[candidate.Path] {
			return
		}
		seenCandidates[candidate.Path] = true
		candidates = append(candidates, *candidate)
	}
	rootForCandidate := func(path, fallback string) string {
		directory := filepath.Dir(path)
		if root, ok := rootCache[directory]; ok {
			return root
		}
		root := findGitRoot(directory)
		if root == "" {
			root = fallback
		}
		rootCache[directory] = root
		return root
	}
	for _, inputPath := range config.Paths {
		absolutePath, err := filepath.Abs(inputPath)
		if err != nil {
			return Result{}, err
		}
		resolvedInputPath := absolutePath
		if resolvedPath, resolveErr := filepath.EvalSymlinks(absolutePath); resolveErr == nil {
			resolvedInputPath = resolvedPath
		}
		info, err := os.Stat(resolvedInputPath)
		if err != nil {
			return Result{}, fmt.Errorf("stat %s: %w", inputPath, err)
		}
		if info.Mode().IsRegular() {
			candidate, reason := candidateFor(resolvedInputPath, workingDirectory, rootForCandidate(resolvedInputPath, inputGitRoot(resolvedInputPath, false, gitRoot)), config)
			if candidate != nil {
				addCandidate(candidate)
			} else {
				skipped++
				if config.Debug {
					diagnostics = append(diagnostics, Diagnostic{Path: displayPath(absolutePath, workingDirectory), Reason: reason})
				}
			}
			continue
		}
		walkPath := absolutePath
		if resolvedPath, resolveErr := filepath.EvalSymlinks(absolutePath); resolveErr == nil {
			walkPath = resolvedPath
		}
		inputRoot := inputGitRoot(walkPath, true, gitRoot)
		err = filepath.WalkDir(walkPath, func(path string, entry os.DirEntry, walkErr error) error {
			if walkErr != nil {
				return walkErr
			}
			if entry.IsDir() {
				if fixedDirectories[entry.Name()] {
					if config.Debug {
						diagnostics = append(diagnostics, Diagnostic{Path: displayPath(path, workingDirectory), Reason: "excluded directory"})
					}
					skipped++
					return filepath.SkipDir
				}
				return nil
			}
			if !entry.Type().IsRegular() {
				return nil
			}
			candidate, reason := candidateFor(path, workingDirectory, rootForCandidate(path, inputRoot), config)
			if candidate != nil {
				addCandidate(candidate)
			} else {
				skipped++
				if config.Debug {
					diagnostics = append(diagnostics, Diagnostic{Path: displayPath(path, workingDirectory), Reason: reason})
				}
			}
			return nil
		})
		if err != nil {
			return Result{}, fmt.Errorf("walk %s: %w", inputPath, err)
		}
	}
	ignored := make(map[string]bool)
	byRoot := make(map[string][]Candidate)
	for _, candidate := range candidates {
		if candidate.gitRoot != "" {
			byRoot[candidate.gitRoot] = append(byRoot[candidate.gitRoot], candidate)
		}
	}
	for root, rootCandidates := range byRoot {
		rootIgnored, ignoreErr := ignoredPaths(rootCandidates, root)
		if ignoreErr != nil {
			return Result{}, ignoreErr
		}
		for path := range rootIgnored {
			ignored[path] = true
		}
	}
	filtered := candidates[:0]
	for _, candidate := range candidates {
		if ignored[candidate.Path] {
			skipped++
			if config.Debug {
				diagnostics = append(diagnostics, Diagnostic{Path: candidate.RelPath, Reason: "matched .gitignore"})
			}
			continue
		}
		filtered = append(filtered, candidate)
	}
	return Result{Candidates: filtered, Skipped: skipped, Debug: diagnostics}, nil
}

func candidateFor(path, workingDirectory, gitRoot string, config Config) (*Candidate, string) {
	if resolvedPath, err := filepath.EvalSymlinks(path); err == nil {
		path = resolvedPath
	}
	language, ok := languages.Lookup(path)
	if !ok {
		return nil, "unsupported file type"
	}
	if len(config.Languages) > 0 && !config.Languages[language] {
		return nil, "language filter"
	}
	relativePath := displayPath(path, workingDirectory)
	if matchesAny(config.Excludes, relativePath) {
		return nil, "exclude glob"
	}
	if len(config.Includes) > 0 && !matchesAny(config.Includes, relativePath) {
		return nil, "include glob"
	}
	return &Candidate{Path: path, RelPath: relativePath, Language: language, gitRoot: gitRoot}, ""
}

func inputGitRoot(path string, directory bool, fallback string) string {
	searchPath := path
	if !directory {
		searchPath = filepath.Dir(path)
	}
	if root := findGitRoot(searchPath); root != "" {
		return root
	}
	return fallback
}

func findGitRoot(path string) string {
	command := exec.Command("git", "-C", path, "rev-parse", "--show-toplevel")
	output, err := command.Output()
	if err != nil {
		return ""
	}
	root, err := filepath.Abs(strings.TrimSpace(string(output)))
	if err != nil {
		return ""
	}
	return root
}

func ignoredPaths(candidates []Candidate, gitRoot string) (map[string]bool, error) {
	ignored := make(map[string]bool)
	if gitRoot == "" || len(candidates) == 0 {
		return ignored, nil
	}
	var input bytes.Buffer
	paths := make([]string, 0, len(candidates))
	relativeToAbsolute := make(map[string]string, len(candidates))
	resolvedRoot, err := filepath.EvalSymlinks(gitRoot)
	if err != nil {
		resolvedRoot = gitRoot
	}
	for _, candidate := range candidates {
		resolvedPath, resolveErr := filepath.EvalSymlinks(candidate.Path)
		if resolveErr != nil {
			resolvedPath = candidate.Path
		}
		relative, err := filepath.Rel(resolvedRoot, resolvedPath)
		if err != nil || strings.HasPrefix(relative, ".."+string(filepath.Separator)) {
			continue
		}
		relative = filepath.ToSlash(relative)
		paths = append(paths, relative)
		relativeToAbsolute[relative] = candidate.Path
		input.WriteString(relative)
		input.WriteByte(0)
	}
	if len(paths) == 0 {
		return ignored, nil
	}
	command := exec.Command("git", "-C", gitRoot, "check-ignore", "--stdin", "--no-index", "-z")
	command.Stdin = &input
	output, err := command.Output()
	if err != nil {
		if exitError, ok := err.(*exec.ExitError); ok && exitError.ExitCode() == 1 {
			return ignored, nil
		}
		return nil, fmt.Errorf("check .gitignore: %w", err)
	}
	for _, path := range bytes.Split(output, []byte{0}) {
		if len(path) == 0 {
			continue
		}
		if absolute, ok := relativeToAbsolute[string(path)]; ok {
			ignored[absolute] = true
		}
	}
	return ignored, nil
}

func displayPath(path, workingDirectory string) string {
	relative, err := filepath.Rel(workingDirectory, path)
	if err != nil {
		return filepath.ToSlash(path)
	}
	return filepath.ToSlash(relative)
}

func matchesAny(patterns []string, path string) bool {
	for _, pattern := range patterns {
		for _, item := range strings.Split(pattern, ",") {
			if globMatch(strings.TrimSpace(item), path) {
				return true
			}
		}
	}
	return false
}

func globMatch(pattern, path string) bool {
	pattern = filepath.ToSlash(pattern)
	path = filepath.ToSlash(path)
	if pattern == "" {
		return false
	}
	if !strings.Contains(pattern, "/") {
		pattern = "**/" + pattern
	}
	expression := "^" + globExpression(pattern) + "$"
	matched, err := regexp.MatchString(expression, path)
	return err == nil && matched
}

func globExpression(pattern string) string {
	var result strings.Builder
	for index := 0; index < len(pattern); index++ {
		switch pattern[index] {
		case '*':
			if index+1 < len(pattern) && pattern[index+1] == '*' {
				index++
				if index+1 < len(pattern) && pattern[index+1] == '/' {
					index++
					result.WriteString("(?:.*/)?")
				} else {
					result.WriteString(".*")
				}
			} else {
				result.WriteString("[^/]*")
			}
		case '?':
			result.WriteString("[^/]")
		default:
			result.WriteString(regexp.QuoteMeta(string(pattern[index])))
		}
	}
	return result.String()
}

File: src/discovery.js
import { promises as fs } from "node:fs";
import { execFile } from "node:child_process";
import path from "node:path";
import { promisify } from "node:util";
import { lookup } from "./languages.js";

const execFileAsync = promisify(execFile);
const skippedDirectories = new Set([".git", "node_modules", "vendor", "dist", "build", "coverage"]);

export function normalizePath(value) { return String(value).replaceAll("\\", "/"); }

export function globToRegExp(pattern) {
  let expression = "^";
  const normalized = normalizePath(pattern);
  for (let index = 0; index < normalized.length; index += 1) {
    const character = normalized[index];
    if (character === "*" && normalized[index + 1] === "*") {
      if (normalized[index + 2] === "/") { expression += "(?:.*/)?"; index += 2; }
      else { expression += ".*"; index += 1; }
    } else if (character === "*") expression += "[^/]*";
    else if (character === "?") expression += "[^/]";
    else expression += /[.+^${}()|[\]\\]/.test(character) ? `\\${character}` : character;
  }
  return new RegExp(`${expression}$`);
}

function matches(relativePath, patterns) { return patterns.length === 0 || patterns.some((pattern) => globToRegExp(pattern).test(relativePath)); }
function excluded(relativePath, patterns) { return patterns.some((pattern) => globToRegExp(pattern).test(relativePath)); }
function display(filePath, cwd) { return normalizePath(path.relative(cwd, filePath) || path.basename(filePath)); }

async function walk(directory, cwd, options, diagnostics) {
  const entries = await fs.readdir(directory, { withFileTypes: true });
  const files = [];
  for (const entry of entries) {
    const fullPath = path.join(directory, entry.name);
    const relative = display(fullPath, cwd);
    if (entry.isSymbolicLink()) { diagnostics.push({ path: relative, reason: "symbolic link" }); continue; }
    if (entry.isDirectory()) {
      if (skippedDirectories.has(entry.name)) diagnostics.push({ path: relative, reason: "fixed directory exclusion" });
      else files.push(...await walk(fullPath, cwd, options, diagnostics));
      continue;
    }
    const language = lookup(fullPath);
    if (!entry.isFile() || !language) { if (entry.isFile()) diagnostics.push({ path: relative, reason: "unsupported file" }); continue; }
    if (!matches(relative, options.includes || [])) { diagnostics.push({ path: relative, reason: "not included" }); continue; }
    if (excluded(relative, options.excludes || [])) { diagnostics.push({ path: relative, reason: "excluded" }); continue; }
    if (options.languages?.size && !options.languages.has(language)) { diagnostics.push({ path: relative, reason: "language filter" }); continue; }
    files.push({ path: fullPath, relative, language });
  }
  return files;
}

async function tracked(cwd) {
  const { stdout } = await execFileAsync("git", ["-C", cwd, "ls-files", "--cached", "--others", "--exclude-standard", "-z"], { encoding: "utf8" });
  return stdout.split("\0").filter(Boolean).map((file) => path.resolve(cwd, file));
}

export async function discover(inputPaths = [], options = {}) {
  const cwd = path.resolve(options.cwd || process.cwd());
  const diagnostics = [];
  const candidates = [];
  const paths = inputPaths.length ? inputPaths : null;
  if (!paths) {
    try {
      for (const fullPath of await tracked(cwd)) {
        const language = lookup(fullPath); const relative = display(fullPath, cwd);
        if (!language || relative.split("/").some((part) => skippedDirectories.has(part))) continue;
        if (matches(relative, options.includes || []) && !excluded(relative, options.excludes || []) && (!options.languages?.size || options.languages.has(language))) candidates.push({ path: fullPath, relative, language });
        else diagnostics.push({ path: relative, reason: "selection filter" });
      }
    } catch { candidates.push(...await walk(cwd, cwd, options, diagnostics)); }
  } else {
    for (const input of paths) {
      const fullPath = path.resolve(cwd, input);
      try {
        const stats = await fs.lstat(fullPath);
        if (stats.isDirectory()) candidates.push(...await walk(fullPath, cwd, options, diagnostics));
        else if (stats.isFile()) {
          const language = lookup(fullPath); const relative = display(fullPath, cwd);
          if (!language) diagnostics.push({ path: relative, reason: "unsupported file" });
          else if (!matches(relative, options.includes || [])) diagnostics.push({ path: relative, reason: "not included" });
          else if (excluded(relative, options.excludes || [])) diagnostics.push({ path: relative, reason: "excluded" });
          else if (options.languages?.size && !options.languages.has(language)) diagnostics.push({ path: relative, reason: "language filter" });
          else candidates.push({ path: fullPath, relative, language });
        }
      } catch (error) { throw new Error(`${input}: ${error.message}`); }
    }
  }
  const unique = [...new Map(candidates.map((candidate) => [path.resolve(candidate.path), candidate])).values()];
  unique.sort((left, right) => left.relative.localeCompare(right.relative));
  return { candidates: unique, diagnostics, skipped: diagnostics.length };
}

export { skippedDirectories };
